source/common/network/win32_socket_handle_impl.cc

```diff
@@ -84,5 +121,74 @@ void Win32SocketHandleImpl::reEnableEventBasedOnIOResult(const Api::IoCallUint64
   }
 }
 
+Api::IoCallUint64Result Win32SocketHandleImpl::drainToPeekBuffer() {
+  while (true) {
+    Buffer::OwnedImpl read_buffer;
+    Buffer::Reservation reservation = read_buffer.reserveForRead();
+    Api::IoCallUint64Result result = IoSocketHandleImpl::readv(
+        reservation.length(), reservation.slices(), reservation.numSlices());
+    uint64_t bytes_to_commit = result.ok() ? result.return_value_ : 0;
+    reservation.commit(bytes_to_commit);
+    peek_buffer_->add(read_buffer);
+    if (!result.ok() || bytes_to_commit == 0) {
+      return result;
+    }
+  }
+}
+
+Api::IoCallUint64Result Win32SocketHandleImpl::readFromPeekBuffer(void* buffer, size_t length) {
+  uint64_t copy_size = std::min(peek_buffer_->length(), static_cast<uint64_t>(length));
+  peek_buffer_->copyOut(0, copy_size, buffer);
+  peek_buffer_->drain(copy_size);
+  return Api::IoCallUint64Result(copy_size, Api::IoErrorPtr(nullptr, [](Api::IoError*) {}));
+}
+
+Api::IoCallUint64Result Win32SocketHandleImpl::readvFromPeekBuffer(uint64_t max_length,
+                                                                   Buffer::RawSlice* slices,
+                                                                   uint64_t num_slice) {
+  uint64_t total_length_to_read = std::min(max_length, peek_buffer_->length());
+  uint64_t num_slices_to_read = 0;
+  uint64_t num_bytes_to_read = 0;
+  for (; num_slices_to_read < num_slice && num_bytes_to_read < total_length_to_read;
+       num_slices_to_read++) {
+    auto length_to_copy = std::min(static_cast<uint64_t>(slices[num_slices_to_read].len_),
+                                   total_length_to_read - num_bytes_to_read);
+    peek_buffer_->copyOut(num_bytes_to_read, length_to_copy, slices[num_slices_to_read].mem_);
+    num_bytes_to_read += length_to_copy;
+  }
+  peek_buffer_->drain(num_bytes_to_read);
+  return Api::IoCallUint64Result(num_bytes_to_read, Api::IoErrorPtr(nullptr, [](Api::IoError*) {}));
+}
+
+Api::IoCallUint64Result Win32SocketHandleImpl::readFromPeekBuffer(Buffer::Instance& buffer,
+                                                                  size_t length) {
+  auto lenght_to_move = std::min(peek_buffer_->length(), static_cast<uint64_t>(length));
+  buffer.move(*peek_buffer_, lenght_to_move);
+  peek_buffer_->drain(lenght_to_move);
```

Comment:
it should be my mistake, after move, we needn't the drain.

Reply:
fixed in 1534dcd520695bcb1e871799461681d9d73f0f22